Add --user-data-dir CLI flag and propose renaming support_dir to data_dir

mkungla opened the pull request:
This PR introduces support for a `--user-data-dir` CLI flag to override Zed's data directory and proposes renaming `support_dir` to `data_dir` for better cross-platform clarity. It builds on the discussion in #25349 about custom data directories, aiming to provide a flexible cross-platform solution.

### Changes

The PR is split into two commits:
1. **[feat(cli): add --user-data-dir to override data directory](https://github.com/zed-industries/zed/pull/26886/commits/28e8889105847401e783d1739722d0998459fe5a)**
2. **[refactor(paths): rename support_dir to data_dir for cross-platform clarity](https://github.com/zed-industries/zed/pull/26886/commits/affd2fc606b39af1b25432a688a9006229a8fc3a)**


### Context
Inspired by the need for custom data directories discussed in #25349, this PR provides an immediate implementation in the first commit, while the second commit suggests a naming improvement for broader appeal. @mikayla-maki, I’d appreciate your feedback, especially on the rename proposal, given your involvement in the original discussion!

### Testing
- `cargo build `
- `./target/debug/zed --user-data-dir ~/custom-data-dir`

Release Notes:
- Added --user-data-dir CLI flag

mkungla committed:
feat(cli): add --user-data-dir to override data directory

Introduces a new CLI flag `--user-data-dir` to allow users to specify a custom directory for Zed's data (e.g., database, extensions, logs). This is implemented by:
- Adding the flag to `crates/cli/src/main.rs` and `crates/zed/src/main.rs`.
- Setting `CUSTOM_DATA_DIR` via `paths::set_custom_data_dir` in both CLI and editor processes.
- Passing the flag via IPC in `CliRequest::Open` for background mode consistency.

mkungla committed:
refactor(paths): rename support_dir to data_dir for cross-platform clarity

Renames `support_dir` to `data_dir` to better reflect its role as the root directory for
Zed's user data across all platforms. The original name was macOS-specific (aligned with "Application Support"),
which felt less intuitive on Linux (~/.local/share/zed) and Windows (%LOCALAPPDATA%\Zed). This change:
- Updates `crates/paths/src/paths.rs` to use `data_dir` and `DATA_DIR` static.
- Adjusts dependent functions (e.g., `database_dir`, `extensions_dir`) to call `data_dir`.
- Updates callers in `crates/cli/src/main.rs` and other crates as needed.

mkungla committed:
fix: update run_foreground args win, mac

Signed-off-by: Marko Kungla <marko.kungla@gmail.com>

mikayla-maki commented on crates/paths/src/paths.rs:
```diff
@@ -5,15 +5,45 @@ use std::sync::OnceLock;
+            PathBuf::from(dir)
+        } else if dir.starts_with("./") || dir.starts_with("../") || (!dir.starts_with("/") && !cfg!(target_os = "windows")) {
+            // Relative path on any OS (Linux/macOS: no leading "/", Windows: no drive letter or UNC)
+            std::env::current_dir()
```
I don't think we need any of these branches, `std::create_dir_all` calls `libc::mkdir` in a loop, and that supports relative paths: https://man7.org/linux/man-pages/man2/mkdir.2.html

mkungla replied:
updated

mikayla-maki commented on crates/zed/src/zed/open_listener.rs:
```diff
@@ -261,7 +261,12 @@ pub async fn handle_cli_connection(
+                user_data_dir,
             } => {
+                if let Some(dir) = user_data_dir {
+                    paths::set_custom_data_dir(&dir);
```
I don't think we need this, as you're already passing it via CLI arguments and set_custom_data_dir() should not be called after `init_paths`. In fact, can we make `set_custom_data_dir()` panic if you call it after calling `init_paths`?

mkungla replied:
updated

mkungla committed:
refactor(paths): simplify set_custom_data_dir per review feedback

Removes unnecessary path resolution branches in `set_custom_data_dir` as
suggested by @mikayla-maki in PR review. `std::fs::create_dir_all` already handles
relative and absolute paths correctly, so we now just use `PathBuf::from(dir)` directly.

mkungla committed:
refactor(paths): enforce early set_custom_data_dir and remove redundant IPC call

Per #26886s review feedback:
- Removed `set_custom_data_dir` from `handle_cli_connection` in `open_listener.rs`, as it’s already set via CLI args and early in `main()`.
- Updated `set_custom_data_dir` in `paths.rs` to panic if called after `data_dir` is initialized, ensuring it’s only set before path initialization.

mkungla committed:
fix(windows): add missing user_data_dir field in windows_only_instance

Resolves a Windows CI compilation error by adding the now required `user_data_dir`
field to `CliRequest::Open` in `windows_only_instance.rs`. Sets it to `None` since
this module forwards paths/URLs without needing to override the data directory,
which is handled by the first instance’s CLI args.

mkungla committed:
refactor(paths): refine CUSTOM_DATA_DIR and CURRENT_DATA_DIR for consistent initialization

Updates `paths.rs` to improve the `--user-data-dir` feature’s path handling:
- Reintroduces `CUSTOM_DATA_DIR` with `CURRENT_DATA_DIR` to separate custom overrides from resolved data dir, fixing recursion in `data_dir()`.
- Ensures `set_custom_data_dir()` sets `CUSTOM_DATA_DIR` early and panics if `CURRENT_DATA_DIR` or `CONFIG_DIR` is initialized.
- Preserves platform defaults for `config_dir()` and `data_dir()` when no custom dir is set, using `CUSTOM_DATA_DIR` otherwise.
- Moves `CONFIG_DIR` to module scope for visibility, matching other statics.
Prevents inconsistent paths and aligns with early initialization requirements.

Signed-off-by: Marko Kungla <marko.kungla@gmail.com>

mikayla-maki commented on crates/paths/src/paths.rs:
```diff
@@ -5,61 +5,84 @@ use std::sync::OnceLock;
+static CUSTOM_DATA_DIR: OnceLock<PathBuf> = OnceLock::new();
+
+// Resolved data directory, combining custom or platform defaults, set once.
+static CURRENT_DATA_DIR: OnceLock<PathBuf> = OnceLock::new();
```
Why do we need this third static? I would have thought the check would be something like:

```rs
pub fn set_custom_data_dir(dir: &str) -> &'static PathBuf {
  if CUSTOM_DATA_DIR.get().is_some() {
   panic!();
  } 
  // Set custom data dir
}
```

And otherwise, it would be the same as it was earlier in the PR

mkungla replied:
I added `CURRENT_DATA_DIR` and `CONFIG_DIR` as separate statics to ensure `set_custom_data_dir` can only be called before any directory (data or config) is initialized. Checking only `CUSTOM_DATA_DIR.get().is_some()` would allow setting a custom directory after `data_dir()` or `config_dir()` has been resolved, which could lead to inconsistent paths during runtime. If you are okay with that possibility, I can simplify it as suggested - let me know what you think!

mikayla-maki commented on crates/paths/src/paths.rs:
```diff
@@ -5,61 +5,84 @@ use std::sync::OnceLock;
+
 /// Returns the path to the configuration directory used by Zed.
 pub fn config_dir() -> &'static PathBuf {
-    static CONFIG_DIR: OnceLock<PathBuf> = OnceLock::new();
```
I don't think there's any reason to move these out of their functions?

mkungla replied:
I moved `CONFIG_DIR` and `CURRENT_DATA_DIR` out of the functions so `set_custom_data_dir` can check their initialization state and panic if they’ve already been set. Keeping them inside would prevent this check. They’re still private to the module (not `pub`), so it doesn’t affect external access. I could refactor to avoid this (e.g., by tracking initialization differently), but I think this approach is cleanest for now. Happy to adjust if you have a better suggestion!

mikayla-maki replied:
Ahh, I see. Thank you!

mkungla committed:
style(cli): improve function signature formatting and documentation

- Fix formatting of run_foreground function signature in InstalledApp trait
- Improve documentation for user_data_dir argument with platform-specific details
- Follow Rust style guidelines for multi-line function signa

Signed-off-by: Marko Kungla <marko.kungla@gmail.com>

mikayla-maki committed:
fmt

mkungla commented:
@mikayla-maki, just wanted to say a big thank you for your guidance on this feature. My initial approach in #25349 was off, but your feedback helped me understand the right way to implement it. I'm glad we could get this `--user-data-dir` flag merged, and I appreciate your support throughout the process. Looking forward to contributing more to Zed!

mikayla-maki commented:
@mkungla Good working with you as well! Thanks for being patient while my availability for OS work was low :D

teohhanhui commented:
@mkungla It seems to be a [bug](https://github.com/zed-industries/zed/issues/28944) introduced in this PR.

mkungla commented:
> @mkungla It seems to be a [bug](https://github.com/zed-industries/zed/issues/28944) introduced in this PR.

pushd pr which should fix it #28952
